rethinkdb/datadog_checks/rethinkdb/document_db/query.py

```diff
@@ -0,0 +1,118 @@
+        self.groups = [] if groups is None else groups
+
+    def _make_metric_from_spec(self, document, spec, tags, logger):
+        # type: (Any, MetricSpec, List[str], CheckLoggingAdapter) -> Metric
+        logger.trace('make_metric_from_spec %r', spec)
+
+        path = spec['path']
+        name = spec.get('name', path)
+        value = lookup_dotted(document, path=path)
+
+        if 'modifier' in spec and spec['modifier'] is not None:
+            value = self._modify(value, modifier=spec['modifier'], logger=logger)
+
+        if not isinstance(value, (int, float)):  # pragma: no cover
+            raise RuntimeError('Expected float or int, got {!r} of type {}', value, type(value))
+
+        name = dotted_join((self.prefix, name))
+
+        return {'type': spec['type'], 'name': name, 'value': value, 'tags': tags}
+
+    def _make_metrics_from_enumeration(self, document, enumeration, tags, logger):
+        # type: (Any, Enumeration, List[str], CheckLoggingAdapter) -> Iterator[Metric]
+        logger.trace('make_metrics_from_enumeration enumeration=%r', enumeration)
+
+        values = lookup_dotted(document, path=enumeration['path'])  # type: Sequence
+
+        for index, value in enumerate(values):
+            item_tags = tags + ['{}:{}'.format(enumeration['index_tag'], index)]
+
+            for spec in enumeration['metrics']:
+                spec = {
+                    'type': spec['type'],
+                    'name': dotted_join((enumeration['path'], spec['path'])),
+                    'path': spec['path'],
+                    'modifier': spec.get('modifier'),
+                }
+                yield self._make_metric_from_spec(value, spec, tags=item_tags, logger=logger)
+
+    def _make_metrics_from_group(self, document, group, tags, logger):
+        # type: (Any, Group, List[str], CheckLoggingAdapter) -> Iterator[Metric]
+        logger.trace('make_metrics_from_group group=%r', group)
+
+        mapping = lookup_dotted(document, path=group['path'])  # type: Mapping
+
+        for key in mapping:
+            item_tags = tags + ['{}:{}'.format(group['key_tag'], key)]
+            spec = {
+                'type': group['value_metric_type'],
+                'name': group['path'],
+                'path': key,
+            }  # type: MetricSpec
+            yield self._make_metric_from_spec(mapping, spec, tags=item_tags, logger=logger)
+
+    def _modify(self, value, modifier, logger):
+        # type: (Any, Modifier, CheckLoggingAdapter) -> float
+        logger.trace('modify value=%r modifier=%r', value, modifier)
+
+        if modifier == 'total':
+            return len(value)
+
+        if modifier == 'ok_warning':
+            return AgentCheck.OK if value else AgentCheck.WARNING
+
+        if modifier == 'time_elapsed':
+            return to_time_elapsed(value)
+
+        raise RuntimeError('Unknown modifier: {!r}'.format(modifier))  # pragma: no cover
+
+    def run(self, *args, **kwargs):
```

Comment:
Any reason not using the explicit arguments here ?

```suggestion
    def run(self, conn, config, logger):
```

Reply:
The goal is for `DocumentQuery` to not be tied to the RethinkDB check — hence why there are no `conn` and `config` parameters, and args/kwargs are just passed to the `source` operation.

Given this, I can't do `*args, logger, **kwargs` because of Python 2 not supporting keyword-only arguments, so I have to `.pop()` the logger from the kwargs…

Reply:
Worked around this like this (pushing soon) -- as a result `query.run()` now only accepts kwargs (no args anymore, which sounds like a fine constraint):

```python
def run(self, log_debug=(lambda *args, **kwargs: None), **kwargs): ...
```

Reply:
> The goal is for DocumentQuery to not be tied to the RethinkDB check

1. I understand, but seems there is maybe a better way to not have to pass things like `log` and `config`, but it's a minor thing. So up to you.

2. `log_debug=no_op` this looks a bit weird :)

Reply:
1) No, I agree, it does look like there's room for improvements, so open to any ideas. Eg in SQL world we have the `QueryManager` which has access to the `check` object and does the logging (not the queries themselves).

Reply:
After looking into it this morning, I really can't find another way to pass runtime parameters (eg `conn`) to query source operations than just forwarding `**kwargs`.

I pushed a slight refactor of the logging stuff in b472756e6: we now pass an optional `AgentCheck` instance and get the logger from it. If this doesn't sound crazy to you I'd suggest we leave it at it. :-)

Reply:
Thx for the changes.